communication/comms.py

```diff
@@ -0,0 +1,185 @@
+        optional keyward arguments ip_address and port_no.
+
+        Parameters
+        ----------
+        name : str
+            Name of pokemon.
+        nickname : str
+            Nickname of pokemon.
+        pokemon_type : str
+            Type of pokemon.
+        ip_address : str
+            The IP address as the client or server (default 127.0.0.1).
+        port_no : int
+            The port number to communicate on (default 8765).
+        """
+        self.ip_address = ip_address
+        self.port_no = port_no
+        super(Comms, self).__init__(name, nickname, pokemon_type)
+
+    async def msg_tx(self, message, writer):
+        """Co-routine writes message to socket.
+
+        Parameters
+        ----------
+        message : str
+            The message to write to socket.
+        writer :  StreamWriter object
+            Object instance of StreamWriter class for writing to socket.
+        """
+        print(f"Sent: {message!r}")
+        writer.write(message.encode())
+        await writer.drain()
+
+    async def list_tasks(self, writer):  # noqa: D301
+        """Co-routine create a random number of tasks for writing messages to socket.
+
+        Each message ends with a \n. Multiple messages are sent as a group
+        of messages, and gets terminated with 'EOM' to indicate the end of transmission.
+
+        Parameters
+        ----------
+        writer :  StreamWriter object
+            Object instance of StreamWriter class for writing to socket.
+        """
+        message = self.name + ".\n"
+        tasks = []
+        random_no = random.randint(1, 5)
+        print(f"Random number generated = {random_no}")
+        for i in range(random_no):
+            tasks.append(self.msg_tx(message, writer))
+        await asyncio.gather(*tasks)
+        await self.msg_tx("EOM\n", writer)
+
+    async def read_data(self, reader):
+        """Co-routine reads from socket.
+
+        The reading of data is carried out in a while loop. All received messages are stored in list.
+        Breaks out of while loop if a message is received containing 'EOM' or 'last message'
+
+        Parameters
+        ----------
+        reader :  StreamReader object
+            Object instance of StreamReader class for reading from socket.
+
+        Returns
+        -------
+        len(msg_list) : tuple int
+            Integer value indicating number of messages received and appended to list.
+        end : tuple bool
+            True or False indicating if connection should be ended.
+        """
+        msg_list = []
+        end = False
+        while True:
+            data = await reader.readline()
+            print(f"Received: {data.decode()!r}")
+            msg_list.append(data)
+            if data.decode() == "EOM\n":
+                print("EOM received and breaking out.")
+                print(f"Number of messages received = {len(msg_list) - 1}")
```

Comment:
You wouldn't need to do the `-1` if you moved the `msg_list.append(data)` after the end of the `if` statement.